File: packages/oidc/android/src/main/java/com/pingidentity/rnoidc/OidcConfigParser.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rnoidc

import com.facebook.react.bridge.ReadableMap
import com.reactnativepingidentity.core.utils.readStringMap
import com.reactnativepingidentity.core.utils.requireString
import com.reactnativepingidentity.core.utils.requireStringArray

/**
 * Parsed configuration payload supplied by the JS bridge.
 */
internal data class OidcClientPayload(
  val clientId: String,
  val discoveryEndpoint: String,
  val openId: OpenIdPayload?,
  val redirectUri: String,
  val scopes: List<String>,
  val storageId: String?,
  val loggerId: String?,
  val acrValues: String?,
  val signOutRedirectUri: String?,
  val state: String?,
  val nonce: String?,
  val uiLocales: String?,
  val refreshThreshold: Long?,
  val loginHint: String?,
  val display: String?,
  val prompt: String?,
  val additionalParameters: Map<String, String>
)

/**
 * Optional OpenID configuration override supplied by JS.
 */
internal data class OpenIdPayload(
  val authorizationEndpoint: String,
  val tokenEndpoint: String,
  val userinfoEndpoint: String,
  val endSessionEndpoint: String?,
  val pingEndIdpSessionEndpoint: String?,
  val revocationEndpoint: String?
)

/**
 * Parses JS configuration maps into strongly-typed Kotlin payloads.
 */
internal object OidcConfigParser {

  /**
   * Parse and validate the JS configuration payload for an OIDC client.
   *
   * @throws IllegalArgumentException when required fields are missing
   */
  fun parseClientConfig(config: ReadableMap): OidcClientPayload {
    val clientId = requireString(config, "clientId")
    val openId = parseOpenId(config)
    val discoveryEndpoint = requireString(config, "discoveryEndpoint")
    val redirectUri = requireString(config, "redirectUri")
    val scopes = requireStringArray(config, "scopes")

    return OidcClientPayload(
      clientId = clientId,
      discoveryEndpoint = discoveryEndpoint,
      openId = openId,
      redirectUri = redirectUri,
      scopes = scopes,
      storageId = if (config.hasKey("storageId")) config.getString("storageId") else null,
      loggerId = if (config.hasKey("loggerId")) config.getString("loggerId") else null,
      acrValues = if (config.hasKey("acrValues")) config.getString("acrValues") else null,
      signOutRedirectUri = if (config.hasKey("signOutRedirectUri")) {
        config.getString("signOutRedirectUri")
      } else {
        null
      },
      state = if (config.hasKey("state")) config.getString("state") else null,
      nonce = if (config.hasKey("nonce")) config.getString("nonce") else null,
      uiLocales = if (config.hasKey("uiLocales")) config.getString("uiLocales") else null,
      refreshThreshold = if (config.hasKey("refreshThreshold")) {
        config.getDouble("refreshThreshold").toLong()
      } else {
        null
      },
      loginHint = if (config.hasKey("loginHint")) config.getString("loginHint") else null,
      display = if (config.hasKey("display")) config.getString("display") else null,
      prompt = if (config.hasKey("prompt")) config.getString("prompt") else null,
      additionalParameters = if (config.hasKey("additionalParameters")) {
        readStringMap(config.getMap("additionalParameters"))
      } else {
        emptyMap()
      }
    )
  }

  private fun parseOpenId(config: ReadableMap): OpenIdPayload? {
    if (!config.hasKey("openId")) {
      return null
    }
    val openIdMap = config.getMap("openId") ?: return null
    return OpenIdPayload(
      authorizationEndpoint = requireString(openIdMap, "authorizationEndpoint"),
      tokenEndpoint = requireString(openIdMap, "tokenEndpoint"),
      userinfoEndpoint = requireString(openIdMap, "userinfoEndpoint"),
      endSessionEndpoint = if (openIdMap.hasKey("endSessionEndpoint")) {
        openIdMap.getString("endSessionEndpoint")
      } else {
        null
      },
      pingEndIdpSessionEndpoint = if (openIdMap.hasKey("pingEndIdpSessionEndpoint")) {
        openIdMap.getString("pingEndIdpSessionEndpoint")
      } else {
        null
      },
      revocationEndpoint = if (openIdMap.hasKey("revocationEndpoint")) {
        openIdMap.getString("revocationEndpoint")
      } else {
        null
      }
    )
  }

  /**
   * Build authorization parameter overrides from JS options.
   *
   * @param options JS options map
   * @return Normalized authorization parameters
   */
  fun buildAuthorizeParams(options: ReadableMap): MutableMap<String, String> {
    val params = mutableMapOf<String, String>()
    options.getString("acrValues")?.let { params["acr_values"] = it }
    options.getString("state")?.let { params["state"] = it }
    options.getString("nonce")?.let { params["nonce"] = it }
    options.getString("uiLocales")?.let { params["ui_locales"] = it }
    options.getString("loginHint")?.let { params["login_hint"] = it }
    options.getString("display")?.let { params["display"] = it }
    options.getString("prompt")?.let { params["prompt"] = it }
    val additional = if (options.hasKey("additionalParameters")) {
      readStringMap(options.getMap("additionalParameters"))
    } else {
      emptyMap()
    }
    params.putAll(additional)
    return params
  }
}

File: packages/oidc/ios/OidcConfigParser.swift
//
//  OidcConfigParser.swift
//  RNPingOidc
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import RNPingCore

/// Helpers for parsing JS configuration payloads into native structs.
enum OidcConfigParser {
  /// Parse a JS dictionary into a payload object.
  ///
  /// - Parameter config: Raw JS configuration dictionary.
  /// - Returns: Parsed OIDC client payload.
  /// - Throws: NSError when required configuration is missing.
  static func parseClientConfig(_ config: NSDictionary) throws -> OidcClientPayload {
    let clientId = try ReadableMapUtils.requireString(config, key: "clientId")
    let discoveryEndpoint = config["discoveryEndpoint"] as? String
    let openIdPayload = parseOpenId(config["openId"])
    let iosConfig = config["ios"] as? NSDictionary
    if (discoveryEndpoint == nil || discoveryEndpoint?.isEmpty == true), openIdPayload == nil {
      throw NSError(domain: "RNPingOidc", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing discoveryEndpoint or openId"])
    }
    let redirectUri = try ReadableMapUtils.requireString(config, key: "redirectUri")
    let scopes = try ReadableMapUtils.requireStringArray(config, key: "scopes")

    return OidcClientPayload(
      clientId: clientId,
      discoveryEndpoint: discoveryEndpoint,
      openId: openIdPayload,
      redirectUri: redirectUri,
      scopes: scopes,
      storageId: config["storageId"] as? String,
      loggerId: config["loggerId"] as? String,
      browserType: iosConfig?["browserType"] as? String,
      browserMode: iosConfig?["browserMode"] as? String,
      acrValues: config["acrValues"] as? String,
      signOutRedirectUri: config["signOutRedirectUri"] as? String,
      state: config["state"] as? String,
      nonce: config["nonce"] as? String,
      uiLocales: config["uiLocales"] as? String,
      refreshThreshold: parseRefreshThreshold(config["refreshThreshold"]),
      loginHint: config["loginHint"] as? String,
      display: config["display"] as? String,
      prompt: config["prompt"] as? String,
      additionalParameters: ReadableMapUtils.readStringMap(config["additionalParameters"] as? NSDictionary)
    )
  }

  /// Parse an OpenID override payload from JS.
  ///
  /// - Parameter value: Raw OpenID override payload.
  /// - Returns: Parsed OpenID payload or nil when invalid.
  static func parseOpenId(_ value: Any?) -> OpenIdPayload? {
    guard let map = value as? NSDictionary else {
      return nil
    }
    guard let authorizationEndpoint = map["authorizationEndpoint"] as? String, !authorizationEndpoint.isEmpty,
          let tokenEndpoint = map["tokenEndpoint"] as? String, !tokenEndpoint.isEmpty,
          let userinfoEndpoint = map["userinfoEndpoint"] as? String, !userinfoEndpoint.isEmpty else {
      return nil
    }
    return OpenIdPayload(
      authorizationEndpoint: authorizationEndpoint,
      tokenEndpoint: tokenEndpoint,
      userinfoEndpoint: userinfoEndpoint,
      endSessionEndpoint: map["endSessionEndpoint"] as? String,
      pingEndIdpSessionEndpoint: map["pingEndIdpSessionEndpoint"] as? String,
      revocationEndpoint: map["revocationEndpoint"] as? String
    )
  }

  /// Build per-request authorization parameters.
  ///
  /// - Parameter options: Raw JS authorization options.
  /// - Returns: Normalized authorization parameters.
  static func buildAuthorizeParams(from options: NSDictionary) -> [String: String] {
    var params = [String: String]()
    if let acr = options["acrValues"] as? String { params["acr_values"] = acr }
    if let state = options["state"] as? String { params["state"] = state }
    if let nonce = options["nonce"] as? String { params["nonce"] = nonce }
    if let uiLocales = options["uiLocales"] as? String { params["ui_locales"] = uiLocales }
    if let loginHint = options["loginHint"] as? String { params["login_hint"] = loginHint }
    if let display = options["display"] as? String { params["display"] = display }
    if let prompt = options["prompt"] as? String { params["prompt"] = prompt }
    let additional = ReadableMapUtils.readStringMap(options["additionalParameters"] as? NSDictionary)
    params.merge(additional) { _, new in new }
    return params
  }

  /// Parse refresh threshold from JS.
  ///
  /// - Parameter value: Raw JS refresh threshold.
  /// - Returns: Parsed threshold value, if provided.
  private static func parseRefreshThreshold(_ value: Any?) -> Int64? {
    if let number = value as? NSNumber {
      return number.int64Value
    }
    return nil
  }
}
